debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerSnapshotChangeEventSource.java

```diff
@@ -250,49 +250,40 @@ protected void complete(SnapshotContext<SqlServerPartition, SqlServerOffsetConte
     }
 
     /**
-     * Generate a valid sqlserver query string for the specified table
+     * Generate a valid SQL Server query string for the specified table
      *
      * @param tableId the table to generate a query for
      * @return a valid query string
      */
     @Override
     protected Optional<String> getSnapshotSelect(RelationalSnapshotContext<SqlServerPartition, SqlServerOffsetContext> snapshotContext,
-                                                 TableId tableId) {
-        String modifiedColumns = checkExcludedColumns(tableId);
-        return Optional.of(String.format("SELECT %s FROM [%s].[%s].[%s]",
-                modifiedColumns, tableId.catalog(), tableId.schema(), tableId.table()));
+                                                 TableId tableId, List<String> columns) {
+        String snapshotSelectColumns = columns.stream()
+                .collect(Collectors.joining(", "));
+        return Optional.of(String.format("SELECT %s FROM [%s].[%s].[%s]", snapshotSelectColumns, tableId.catalog(), tableId.schema(), tableId.table()));
     }
 
     @Override
-    protected String enhanceOverriddenSelect(RelationalSnapshotContext<SqlServerPartition, SqlServerOffsetContext> snapshotContext,
-                                             String overriddenSelect, TableId tableId) {
-        String modifiedColumns = checkExcludedColumns(tableId);
-        return overriddenSelect.replaceAll("\\*", modifiedColumns);
-    }
-
-    private String checkExcludedColumns(TableId tableId) {
-        Table table = sqlServerDatabaseSchema.tableFor(tableId);
-        List<String> columnNames = table.retrieveColumnNames().stream()
-                .filter(columnName -> filterChangeTableColumns(tableId, columnName))
-                .filter(columnName -> connectorConfig.getColumnFilter().matches(tableId.catalog(), tableId.schema(), tableId.table(), columnName))
-                .collect(Collectors.toList());
-
-        if (columnNames.isEmpty()) {
-            throw new IllegalArgumentException("Filtered column list for table " + tableId + " is empty");
+    protected String resolveColumnName(TableId tableId, String columnName) {
+        StringBuilder sb = new StringBuilder();
+        if (filterChangeTableColumns(tableId, columnName)) {
```

Comment:
@gunnarmorling Don't think `resolveColumnName()` method can be removed. Here is a checking condition in SQL Server. Wdyt?

Reply:
Ah yes, good call. I think this call to `filterChangeTableColumns()` doesn't actually belong into that resolve method to begin with, but it should be part of the filter application in `getPreparedColumnNames()` (see above, it used to be part of `checkExcludedColumns()`. So you'd have two ways for doing that:

* Simply override `getPreparedColumnNames()` for SQL Server, so it applies the two filters (the normal include/exclude one and this SQL Server specific one based on the change tables)
* You do the main part of the work in `getPreparedColumnNames()` in the base class but then define an additional filter method such as `additionalColumnFilter()` which would only be implemented for SQL Server

The latter would be my preference.

Reply:
@gunnarmorling Addressed this in my last commit https://github.com/debezium/debezium/pull/2626/commits/a18f44f4b564a3be14ee3c9144a15eec06ddd93e, but this broke Postgres and MySQL, hence CI is still running & failing. Without this everything worked well.